### DISCUSSION TITLE
Upgrading QoS level possible?

### DISCUSSION THREAD

**Discussion by Do4bled** [dotnet/MQTTnet#1446] [Q&A] [answered]
I'm not sure if it is normal behavior or that it is a bug because I can't find it in the MQTT specifications. 

According to the specification the QoS level between publisher and broker and broker and subscribers can differ and need to be seen as different processes. 

While testing this I found indeed that when I publish a message from a client using QoS 2, the receiving client that subscribed with QoS 2 will receive a QoS 2 message from the broker and act on that. When I downgrade the receiving client to QoS 1 then the broker sends a message with QoS 1 regardless if the message was published with QoS 2, so that works as specified. 

Now if I turn it around and publish a message from a client using QoS 1 and subscribe with another client using QoS 2 still this client receives a message from the broker using QoS 1. So it seems that the QoS can be downgraded but not upgraded. 

To me this seems odd. Can somebody tell me if this is normal behavior, and if so why?

**Comment by chkr1011** [comment, marked as answer]
From my understanding this is not possible. I also understand the RFC in the way you did and implemented the server accordingly.

To me this makes sense because the overall end to end communication has a certain QoS level. There is not need to use QoS 2 for one receiver and 0 for the sender. The overall situation will be that the messages may getting lost due to level 0 at the sender. Increasing at the receiver will not affect this at all. So upgrading would require to upgrade at the sender as well which is not supported.

**Comment by Do4bled** [comment]
I do not agree on that. it could be possible that a specific client cannot cope with duplicate messages while other clients have no problem with that then you want to use this mechanism (Qos 2) to make sure that this specific client only receive a single message. The publishing client is not aware of all the subscribers and their requirements on how to receive data and could be publishing on QoS 1. 
If the publishing client is designed for a fire and forget (QoS 0) which means that so now and then it is possible that a message got lost. but the specification of the complete chain could still be, that every subscriber (currently active or not) should have all messages that are received by the broker and therefor the subscribers require at least QoS 1.

**Reply by chkr1011** [reply]
From a theoretical standpoint I agree with you but unfortunately the following part of the RFC clearly says that upgrading is not possible.

> If a subscribing Client has been granted maximum QoS 1 for a particular Topic Filter, then a QoS 0 Application Message matching the filter is delivered to the Client at QoS 0. This means that at most one copy of the message is received by the Client. On the other hand, a QoS 2 Message published to the same topic is downgraded by the Server to QoS 1 for delivery to the Client, so that Client might receive duplicate copies of the Message.

I was only trying to find a reason for such a design decision.

**Comment by Do4bled** [comment]
Thank you for the clarification. Going over the Mosquitto broker config file I found the option to activate this, using the mosquitto broker. They also specify that it is against the MQTT spec

The MQTT specification requires that the QoS of a message delivered to a
subscriber is never upgraded to match the QoS of the subscription. Enabling
this option changes this behaviour. If upgrade_outgoing_qos is set true,
messages sent to a subscriber will always match the QoS of its subscription.
This is a non-standard option explicitly disallowed by the spec.
upgrade_outgoing_qos false

**Reply by chkr1011** [reply]
Ah okay. Then please create a feature requests so that we can also adopt this feature in the future.
